[FIXED JENKINS-36232] NPE during SCM polling

ThomasGP opened the pull request:
see [JENKINS-36232](https://issues.jenkins-ci.org/browse/JENKINS-36232) - a NullPointerException introduced in commit 5a2a265fde4bc41210bb40da9cc9522a3bd43f88 / PR #2418 causes the polling e.g. from a git hook to fail.

Added a null check like it is done in the other places SCMDecisionHandler.shouldPoll() is used.

ThomasGP committed:
[FIXED JENKINS-36232] NPE during SCM polling

ashleyfrieze commented:
Please... for the love of all that's holy... please can this bug be fixed. It basically breaks Jenkins!

rpionke commented:
@stephenc

olivergondza commented:
:+1:, even though this is obviously a bug, a test would be nice.

oleg-nenashev commented:
+1 for tests, but they can be done in a separate PR. Hence 👍 
The best way to protect from such issues is to finally fix FindBugs and then enforce it in the PR builder.

The fix will be integrated into the next release. Or do we need an out-of-order release?

ashleyfrieze commented:
From my POV, I've either got to revert my jenkins installation to 2.10, or wait for your next release... so an early next release would fix my "builds don't build automatically any more" problem :)

ThomasGP committed:
add test to trigger the "no veto" code path that has the NPE

ThomasGP commented:
I am all for tests :)

I just added a small test that triggers the NPE. Originally I refrained from testing, since there where some statics and deprecated classes involved. But there you have it, with `@SuppressWarnings("deprecation")` and `PowerMockito.mockStatic`...

oleg-nenashev commented:
@ashleyfrieze I'll ask @kohsuke later today if he is available to do the release

ashleyfrieze commented:
@oleg-nenashev really appreciated... Great work all on here for jumping on this issue. I know it's a one-liner, but it makes a difference!

oleg-nenashev commented on core/src/test/java/jenkins/triggers/SCMTriggerItemTest.java:
```diff
@@ -0,0 +1,38 @@
+
+    @Test
+    @PrepareForTest(SCMDecisionHandler.class)
+    public void noVetoDelegatesPollingToAnSCMedItem() {
```
Nice2have: Add <code>@Issue("JENKINS-36232")</code>

ThomasGP committed:
incorporate remark on using @Issue reference

ThomasGP commented:
added the @Isssue reference as remarked, rebased fix-branch onto current upstream master. I hope both things (and the 3 hour build time) do not interfere with a potential out-of-order release.

oleg-nenashev commented:
They should not. I'll merge the PR once the build finishes successfully

tersmitten commented:
@oleg-nenashev Thanks

oleg-nenashev commented:
@ThomasGP so seems there won't be out-of-order release. As a workaround, you can take the master branch build from CI. Link: https://jenkins.ci.cloudbees.com/job/core/job/jenkins-core/5541/artifact/war/target/jenkins.war

tersmitten commented:
@oleg-nenashev Thanks, build 5541 solved our problems :-)

mpcontracting commented:
Just trialling Jenkins for the first time and hit this bug while testing GitHub auto-build integration - taken me over 5 hours of my Saturday messing around before I found this bug and fixed it with the interim build above. Isn't this the kind of situation that Jenkins is supposed to help solve? This does not inspire confidence people...

ashleyfrieze commented:
I invoke Will Wheaton's first law. 

Perhaps the LTS branch is better for you if you want more guarantees of everything being in perfect condition. 

Perhaps taking your irritation out on the volunteers fixing the problem...

ashleyfrieze commented:
Please see jenkins-ci.org for download instructions.

mpcontracting commented:
OK, this is my last comment. If you google "install jenkins redhat" you end up at this page - https://wiki.jenkins-ci.org/display/JENKINS/Installing+Jenkins+on+Red+Hat+distributions - whilst it does mention the LTS branch, the main installation instruction is the one I followed. It doesn't mention that it's bleeding edge; that it's a weekly build. Maybe this should be updated with a warning that actually it's an unstable branch?

ashleyfrieze commented:
You should definitely report that in a suitable place.
